File: app/src/main/java/com/dpis/module/applist/InstalledAppCatalogCoordinator.java
package com.dpis.module.applist;

import com.dpis.module.DpisConfigStore;

import com.dpis.module.fonts.FontApplyMode;

import com.dpis.module.viewport.ViewportApplyMode;
import com.dpis.module.viewport.ViewportTargetSpec;
import com.dpis.module.viewport.ViewportTargetType;

import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.SystemClock;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public final class InstalledAppCatalogCoordinator {
    public interface Host {
        PackageManager getPackageManager();

        String getSelfPackageName();

    }

    private final Object installedAppCatalogLock = new Object();
    private final Object installedAppCatalogBuildLock = new Object();

    private final Host host;
    private final long installedAppCatalogTtlMs;

    private List<InstalledAppCatalogItem> installedAppCatalog = Collections.emptyList();
    private long installedAppCatalogLoadedAtMs;

    public InstalledAppCatalogCoordinator(Host host,
            long installedAppCatalogTtlMs) {
        this.host = host;
        this.installedAppCatalogTtlMs = installedAppCatalogTtlMs;
    }

    /** Kept for legacy View binders that still notify visible icon rows. */
    public void onIconLoadRequested(String packageName) {
    }

    /** No executor remains after moving icon resolution into the catalog build. */
    public void shutdown() {
    }

    public List<InstalledAppCatalogItem> loadInstalledAppCatalog(
            boolean forceInstalledAppCatalogReload) {
        PackageManager packageManager = host.getPackageManager();
        List<InstalledAppCatalogItem> catalog = getInstalledAppCatalog(
                packageManager,
                host.getSelfPackageName(),
                forceInstalledAppCatalogReload);
        return catalog;
    }

    /**
     * Legacy target pickers display the full list at once. Keep their rows complete while
     * the Compose app workspace resolves icons only for the page it is about to show.
     */
    public List<InstalledAppCatalogItem> loadInstalledAppCatalogWithIcons(
            boolean forceInstalledAppCatalogReload) {
        List<InstalledAppCatalogItem> catalog = loadInstalledAppCatalog(
                forceInstalledAppCatalogReload);
        PackageManager packageManager = host.getPackageManager();
        for (InstalledAppCatalogItem item : catalog) {
            loadItemIcon(packageManager, item);
        }
        return catalog;
    }

    public List<AppListItem> loadInstalledApps(boolean forceInstalledAppCatalogReload,
            DpisConfigStore store,
            Set<String> scopePackages,
            boolean scopeKnown) {
        List<InstalledAppCatalogItem> catalog = loadInstalledAppCatalog(
                forceInstalledAppCatalogReload);
        Set<String> userVisibleConfiguredPackages = userVisibleConfiguredPackages(store);
        List<AppListItem> result = new ArrayList<>(catalog.size());
        for (InstalledAppCatalogItem item : catalog) {
            boolean inScope = scopePackages != null && scopePackages.contains(item.packageName);
            if (userVisibleConfiguredPackages.contains(item.packageName)) {
                // Configured rows retain the complete, compatibility-aware values used when
                // reopening their editor. Most installed packages have no saved DPIS state.
                result.add(createAppListItem(store, scopePackages, scopeKnown,
                        item.label, item.packageName, item.systemApp,
                        item.hyperOsNativeProxyCandidate, true, null));
            } else {
                result.add(createUnconfiguredAppListItem(
                        item.label, item.packageName, inScope, scopeKnown,
                        item.systemApp, item.hyperOsNativeProxyCandidate, true));
            }
        }
        for (String packageName : configuredPackagesMissingFromCatalog(
                userVisibleConfiguredPackages, catalog)) {
            result.add(createAppListItem(store, scopePackages, scopeKnown,
                    packageName, packageName, false, false, false, null));
        }
        return result;
    }

    private static List<String> configuredPackagesMissingFromCatalog(
            Set<String> userVisibleConfiguredPackages,
            List<InstalledAppCatalogItem> catalog) {
        if (userVisibleConfiguredPackages == null || userVisibleConfiguredPackages.isEmpty()) {
            return Collections.emptyList();
        }
        Set<String> installedPackages = new HashSet<>();
        for (InstalledAppCatalogItem item : catalog) {
            installedPackages.add(item.packageName);
        }
        List<String> missing = new ArrayList<>();
        for (String packageName : userVisibleConfiguredPackages) {
            if (!installedPackages.contains(packageName)) {
                missing.add(packageName);
            }
        }
        Collections.sort(missing);
        return missing;
    }

    /**
     * Validates only candidate packages recovered from persistent state. This preserves imported
     * and legacy configuration semantics without running every compatibility lookup for every
     * ordinary installed app.
     */
    static Set<String> userVisibleConfiguredPackages(DpisConfigStore store) {
        if (store == null) {
            return Collections.emptySet();
        }
        Set<String> result = new HashSet<>();
        for (String packageName : store.getConfiguredPackages()) {
            if (store.hasUserVisiblePackageConfig(packageName)) {
                result.add(packageName);
            }
        }
        return result;
    }

    /**
     * An unconfigured row has no persisted package state to decode. Keep its explicit defaults
     * here so a complete app catalogue does not turn into a full config-store scan.
     */
    static AppListItem createUnconfiguredAppListItem(String label,
            String packageName,
            boolean inScope,
            boolean scopeKnown,
            boolean systemApp,
            boolean hyperOsNativeProxyCandidate,
            boolean installed) {
        return new AppListItem(label, packageName,
                inScope, scopeKnown, null, null,
                ViewportApplyMode.OFF, ViewportTargetType.OFF,
                ViewportTargetSpec.off(), null, FontApplyMode.OFF, null,
                false, null, true,
                scopeKnown && inScope, installed,
                systemApp, hyperOsNativeProxyCandidate, null);
    }

    public static AppListItem createAppListItem(DpisConfigStore store,
            Set<String> scopePackages,
            boolean scopeKnown,
            String label,
            String packageName,
            boolean systemApp,
            boolean hyperOsNativeProxyCandidate,
            boolean installed,
            Drawable icon) {
        Integer viewportWidth = store != null
                ? store.getTargetViewportWidthDp(packageName)
                : null;
        Integer viewportScaleMilliPercent = store != null
                ? store.getTargetViewportScaleMilliPercent(packageName)
                : null;
        ViewportTargetSpec viewportTargetSpec = store != null
                ? store.getTargetViewportSpec(packageName)
                : ViewportTargetSpec.off();
        String viewportTargetType = store != null
                ? store.getTargetViewportType(packageName)
                : ViewportTargetType.OFF;
        String viewportMode = store != null
                ? store.getTargetViewportApplyMode(packageName)
                : ViewportApplyMode.OFF;
        Integer fontScalePercent = store != null
                ? store.getTargetFontScalePercent(packageName)
                : null;
        String fontMode = store != null
                ? store.getTargetFontApplyMode(packageName)
                : FontApplyMode.OFF;
        String typefaceId = store != null
                ? store.getTargetTypefaceId(packageName)
                : null;
        boolean appSpecificConfigActive = store != null
                && store.hasTargetAppSpecificConfig(packageName);
        Integer wechatDpi = store != null ? store.getWechatDpi(packageName) : null;
        boolean dpisEnabled = store == null || store.isTargetDpisEnabled(packageName);
        boolean inScope = scopePackages != null && scopePackages.contains(packageName);
        boolean configured = isUserVisibleConfiguredPackage(
                store,
                packageName,
                scopeKnown,
                inScope);
        return new AppListItem(label, packageName,
                inScope, scopeKnown, viewportWidth,
                viewportScaleMilliPercent, viewportMode, viewportTargetType,
                viewportTargetSpec, fontScalePercent, fontMode, typefaceId,
                appSpecificConfigActive, wechatDpi, dpisEnabled, configured, installed,
                systemApp, hyperOsNativeProxyCandidate, icon);
    }

    public static boolean isUserVisibleConfiguredPackage(DpisConfigStore store,
            String packageName,
            boolean scopeKnown,
            boolean inScope) {
        return (scopeKnown && inScope)
                || (store != null && store.hasUserVisiblePackageConfig(packageName));
    }

    private List<InstalledAppCatalogItem> getInstalledAppCatalog(PackageManager packageManager,
            String selfPackageName,
            boolean forceReload) {
        long now = SystemClock.elapsedRealtime();
        synchronized (installedAppCatalogLock) {
            boolean cacheFresh = !installedAppCatalog.isEmpty()
                    && now - installedAppCatalogLoadedAtMs <= installedAppCatalogTtlMs;
            if (!forceReload && cacheFresh) {
                return installedAppCatalog;
            }
        }
        synchronized (installedAppCatalogBuildLock) {
            now = SystemClock.elapsedRealtime();
            synchronized (installedAppCatalogLock) {
                boolean cacheFresh = !installedAppCatalog.isEmpty()
                        && now - installedAppCatalogLoadedAtMs <= installedAppCatalogTtlMs;
                if (!forceReload && cacheFresh) {
                    return installedAppCatalog;
                }
            }

            List<ApplicationInfo> installedApps;
            // The catalogue needs only the package identity, base ApplicationInfo and label.
            // Manifest metadata is substantially more expensive to marshal for every installed
            // package, and the sole consumer (HyperOS proxy eligibility) is resolved later for
            // the one package the user saves or restarts.
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                installedApps = packageManager.getInstalledApplications(
                        PackageManager.ApplicationInfoFlags.of(0L));
            } else {
                installedApps = packageManager.getInstalledApplications(0);
            }

            List<InstalledAppCatalogItem> rebuilt = new ArrayList<>();
            for (ApplicationInfo applicationInfo : installedApps) {
                if (applicationInfo.packageName.equals(selfPackageName)) {
                    continue;
                }
                boolean systemApp = isSystemApp(applicationInfo);
                String label = packageManager.getApplicationLabel(applicationInfo).toString();
                rebuilt.add(new InstalledAppCatalogItem(
                        label,
                        applicationInfo.packageName,
                        systemApp,
                        false,
                        applicationInfo,
                        null));
            }
            rebuilt.sort(Comparator.comparing(
                    (InstalledAppCatalogItem item) -> item.label.toLowerCase(Locale.ROOT))
                    .thenComparing(item -> item.packageName));
            List<InstalledAppCatalogItem> snapshot = Collections.unmodifiableList(rebuilt);
            synchronized (installedAppCatalogLock) {
                installedAppCatalog = snapshot;
                installedAppCatalogLoadedAtMs = now;
                return installedAppCatalog;
            }
        }
    }

    private static Drawable loadApplicationIcon(PackageManager packageManager,
            ApplicationInfo applicationInfo) {
        try {
            return applicationInfo.loadIcon(packageManager);
        } catch (RuntimeException ignored) {
            return null;
        }
    }

    private static Drawable loadItemIcon(PackageManager packageManager,
            InstalledAppCatalogItem item) {
        Drawable cached = item.icon;
        if (cached != null) {
            return cached;
        }
        Drawable loaded = loadApplicationIcon(packageManager, item.applicationInfo);
        if (loaded != null) {
            item.icon = loaded;
        }
        return loaded;
    }

    private static boolean isSystemApp(ApplicationInfo applicationInfo) {
        int flags = applicationInfo.flags;
        return (flags & ApplicationInfo.FLAG_SYSTEM) != 0
                && (flags & ApplicationInfo.FLAG_UPDATED_SYSTEM_APP) == 0;
    }
}

File: app/src/main/java/com/dpis/module/applist/InstalledAppCatalogCoordinator.kt
package com.dpis.module.applist

import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.graphics.drawable.Drawable
import android.os.Build
import com.dpis.module.config.DpisConfigStore
import com.dpis.module.diagnostics.DpisLog
import com.dpis.module.fonts.FontApplyMode
import com.dpis.module.hooks.SystemFrameworkScope
import com.dpis.module.viewport.ViewportApplyMode
import com.dpis.module.viewport.ViewportTargetSpec
import com.dpis.module.viewport.ViewportTargetType
import java.util.Locale

class InstalledAppCatalogCoordinator(
    private val host: Host,
    private val labelStore: InstalledAppCatalogLabelStore? = null,
) {
    interface Host {
        fun getPackageManager(): PackageManager

        fun getSelfPackageName(): String

        fun getCatalogLocaleTag(): String
    }

    class ContextHost(private val context: Context) : Host {
        override fun getPackageManager(): PackageManager = context.packageManager

        override fun getSelfPackageName(): String = context.packageName

        override fun getCatalogLocaleTag(): String {
            val locales = context.resources.configuration.locales
            return catalogLocaleTag(if (locales.isEmpty) null else locales[0])
        }
    }

    private val installedAppCatalogLock = Any()
    private val installedAppCatalogBuildLock = Any()

    private var installedAppCatalog: List<InstalledAppCatalogItem> = emptyList()
    private var catalogInvalidated = true

    /** Kept for legacy View binders that still notify visible icon rows. */
    fun onIconLoadRequested(packageName: String) = Unit

    /** No executor remains after moving icon resolution into the catalog build. */
    fun shutdown() = Unit

    fun invalidate() {
        synchronized(installedAppCatalogLock) {
            catalogInvalidated = true
        }
    }

    fun labelsResolved(): Boolean = synchronized(installedAppCatalogLock) {
        installedAppCatalog.isNotEmpty() && installedAppCatalog.all { it.labelResolved }
    }

    /**
     * Complete catalog with resolved labels. Target pickers keep showing one finished list.
     */
    fun loadInstalledAppCatalog(forceInstalledAppCatalogReload: Boolean): List<InstalledAppCatalogItem> {
        getInstalledAppCatalog(
            host.getPackageManager(),
            host.getSelfPackageName(),
            forceInstalledAppCatalogReload,
        )
        return resolveCatalogLabels(host.getPackageManager())
    }

    /**
     * Legacy target pickers display the full list at once. Keep their rows complete while
     * the Compose app workspace resolves icons only for the page it is about to show.
     */
    fun loadInstalledAppCatalogWithIcons(
        forceInstalledAppCatalogReload: Boolean,
    ): List<InstalledAppCatalogItem> {
        val catalog = loadInstalledAppCatalog(forceInstalledAppCatalogReload)
        val packageManager = host.getPackageManager()
        catalog.forEach { loadItemIcon(packageManager, it) }
        return catalog
    }

    fun loadInstalledApps(
        forceInstalledAppCatalogReload: Boolean,
        store: DpisConfigStore?,
        scopePackages: Set<String>?,
        scopeKnown: Boolean,
    ): List<AppListItem> {
        val catalog = getInstalledAppCatalog(
            host.getPackageManager(),
            host.getSelfPackageName(),
            forceInstalledAppCatalogReload,
        )
        return toAppListItems(catalog, store, scopePackages, scopeKnown, forceInstalledAppCatalogReload)
    }

    fun resolveInstalledAppLabels(
        store: DpisConfigStore?,
        scopePackages: Set<String>?,
        scopeKnown: Boolean,
    ): List<AppListItem> {
        resolveCatalogLabels(host.getPackageManager())
        return loadInstalledApps(false, store, scopePackages, scopeKnown)
    }

    private fun getInstalledAppCatalog(
        packageManager: PackageManager,
        selfPackageName: String,
        forceReload: Boolean,
    ): List<InstalledAppCatalogItem> {
        synchronized(installedAppCatalogLock) {
            if (!forceReload && isCatalogCacheFresh()) {
                DpisLog.i("installed app catalog cache hit: size=${installedAppCatalog.size}")
                return installedAppCatalog
            }
        }
        synchronized(installedAppCatalogBuildLock) {
            synchronized(installedAppCatalogLock) {
                if (!forceReload && isCatalogCacheFresh()) {
                    return installedAppCatalog
                }
            }

            var installedPackages = queryInstalledPackages(packageManager)
            DpisLog.i(
                "installed packages query returned: size=${installedPackages.size}, " +
                    "sdk=${Build.VERSION.SDK_INT}, forceReload=$forceReload",
            )

            if (shouldUseLauncherVisibilityFallback(
                    installedPackages.map { it.packageName },
                    selfPackageName,
                )
            ) {
                installedPackages = mergeLauncherPackages(packageManager, installedPackages)
                DpisLog.i("launcher visibility fallback returned: size=${installedPackages.size}")
            }

            val cachedLabels = labelStore?.load() ?: CatalogLabelCacheSnapshot.EMPTY
            val snapshot = buildCatalogSnapshot(
                installedPackages,
                selfPackageName,
                cachedLabels,
                host.getCatalogLocaleTag(),
            )
            val reusedLabels = snapshot.count { it.labelResolved }
            DpisLog.i(
                "installed app catalog rebuilt: raw=${installedPackages.size}, " +
                    "catalog=${snapshot.size}, reusedLabels=$reusedLabels",
            )
            synchronized(installedAppCatalogLock) {
                installedAppCatalog = snapshot
                catalogInvalidated = false
                return installedAppCatalog
            }
        }
    }

    private fun resolveCatalogLabels(packageManager: PackageManager): List<InstalledAppCatalogItem> {
        synchronized(installedAppCatalogBuildLock) {
            val current = synchronized(installedAppCatalogLock) { installedAppCatalog }
            if (current.isEmpty() || current.all { it.labelResolved }) {
                return current
            }
            val resolved = resolveCatalogItemLabels(
                current,
            ) { applicationInfo ->
                try {
                    packageManager.getApplicationLabel(applicationInfo).toString()
                } catch (_: RuntimeException) {
                    null
                }
            }
            synchronized(installedAppCatalogLock) {
                if (catalogInvalidated) {
                    return current
                }
                installedAppCatalog = resolved
            }
            persistResolvedLabels(resolved)
            return resolved
        }
    }

    private fun persistResolvedLabels(catalog: List<InstalledAppCatalogItem>) {
        val store = labelStore ?: return
        val localeTag = host.getCatalogLocaleTag()
        val merged = mergePersistedLabelRecords(
            store.load(),
            localeTag,
            catalog.mapTo(HashSet()) { it.packageName },
            persistableLabelRecords(catalog),
        )
        store.replace(localeTag, merged)
    }

    private fun isCatalogCacheFresh(): Boolean =
        !catalogInvalidated && installedAppCatalog.isNotEmpty()

    private fun toAppListItems(
        catalog: List<InstalledAppCatalogItem>,
        store: DpisConfigStore?,
        scopePackages: Set<String>?,
        scopeKnown: Boolean,
        forceInstalledAppCatalogReload: Boolean,
    ): List<AppListItem> {
        val configuredPackages = userVisibleConfiguredPackages(
            store,
            scopePackages,
            scopeKnown,
        )
        val result = ArrayList<AppListItem>(catalog.size)
        for (item in catalog) {
            val inScope = scopePackages?.contains(item.packageName) == true
            if (item.packageName in configuredPackages) {
                // Configured rows retain compatibility-aware values used when reopening editors.
                val listItem = createAppListItem(
                    store,
                    scopePackages,
                    scopeKnown,
                    item.label,
                    item.packageName,
                    item.systemApp,
                    item.hyperOsNativeProxyCandidate,
                    true,
                    null,
                )
                listItem.firstInstallTime = item.firstInstallTime
                listItem.lastUpdateTime = item.lastUpdateTime
                result += listItem
            } else {
                val listItem = createUnconfiguredAppListItem(
                    item.label,
                    item.packageName,
                    inScope,
                    scopeKnown,
                    item.systemApp,
                    item.hyperOsNativeProxyCandidate,
                    true,
                )
                listItem.firstInstallTime = item.firstInstallTime
                listItem.lastUpdateTime = item.lastUpdateTime
                result += listItem
            }
        }
        for (packageName in configuredPackagesMissingFromCatalog(configuredPackages, catalog)) {
            result += createAppListItem(
                store,
                scopePackages,
                scopeKnown,
                packageName,
                packageName,
                false,
                false,
                false,
                null,
            )
        }
        DpisLog.i(
            "app list items built: catalog=${catalog.size}, result=${result.size}, " +
                "forceReload=$forceInstalledAppCatalogReload, scopeKnown=$scopeKnown",
        )
        return result
    }

    companion object {
        @JvmStatic
        fun unresolvedCatalogLabel(applicationInfo: ApplicationInfo?, packageName: String): String {
            val nonLocalized = applicationInfo?.nonLocalizedLabel?.toString()?.trim()
            return if (!nonLocalized.isNullOrEmpty()) nonLocalized else packageName
        }

        @JvmStatic
        fun catalogLocaleTag(locale: Locale?): String =
            (locale ?: Locale.getDefault()).toLanguageTag()

        @JvmStatic
        fun buildCatalogSnapshot(
            packages: List<PackageInfo>,
            selfPackageName: String,
            cachedLabels: CatalogLabelCacheSnapshot,
            localeTag: String,
        ): List<InstalledAppCatalogItem> =
            packages.mapNotNull { packageInfo ->
                val unresolved = unresolvedCatalogLabel(
                    packageInfo.applicationInfo,
                    packageInfo.packageName,
                )
                val cached = cachedLabels.resolvedLabel(
                    localeTag,
                    packageInfo.packageName,
                    packageInfo.lastUpdateTime,
                )
                createCatalogItem(
                    packageInfo,
                    selfPackageName,
                    cached ?: unresolved,
                    labelResolved = cached != null,
                )
            }.let(::sortedCatalog)

        @JvmStatic
        fun resolveCatalogItemLabels(
            items: List<InstalledAppCatalogItem>,
            loadLabel: (ApplicationInfo) -> String?,
        ): List<InstalledAppCatalogItem> =
            items.map { item ->
                if (item.labelResolved) {
                    item
                } else {
                    val label = loadLabel(item.applicationInfo)?.trim().orEmpty()
                    if (label.isEmpty()) item else item.withResolvedLabel(label)
                }
            }.let(::sortedCatalog)

        private fun sortedCatalog(items: List<InstalledAppCatalogItem>): List<InstalledAppCatalogItem> =
            items.sortedWith(
                compareBy<InstalledAppCatalogItem> { it.label.lowercase(Locale.ROOT) }
                    .thenBy { it.packageName },
            )

        @JvmStatic
        fun persistableLabelRecords(
            catalog: List<InstalledAppCatalogItem>,
        ): Map<String, CatalogLabelRecord> {
            val records = LinkedHashMap<String, CatalogLabelRecord>()
            for (item in catalog) {
                if (!item.labelResolved) continue
                val label = item.label.trim()
                if (label.isEmpty()) continue
                records[item.packageName] = CatalogLabelRecord(label, item.lastUpdateTime)
            }
            return records
        }

        @JvmStatic
        fun mergePersistedLabelRecords(
            existing: CatalogLabelCacheSnapshot,
            localeTag: String,
            installedPackages: Set<String>,
            updates: Map<String, CatalogLabelRecord>,
        ): Map<String, CatalogLabelRecord> {
            val merged = LinkedHashMap<String, CatalogLabelRecord>()
            if (existing.localeTag == localeTag) {
                existing.records.forEach { (packageName, record) ->
                    if (packageName in installedPackages) {
                        merged[packageName] = record
                    }
                }
            }
            updates.forEach { (packageName, record) ->
                if (packageName in installedPackages) {
                    merged[packageName] = record
                }
            }
            return merged
        }

        @JvmStatic
        fun createCatalogItem(
            packageInfo: PackageInfo,
            selfPackageName: String,
            label: String,
            labelResolved: Boolean,
        ): InstalledAppCatalogItem? {
            val packageName = packageInfo.packageName
            if (packageName.isBlank() || packageName == selfPackageName) return null
            val applicationInfo = packageInfo.applicationInfo
                ?: ApplicationInfo().also { it.packageName = packageName }
            return InstalledAppCatalogItem(
                label,
                packageName,
                isSystemApp(applicationInfo),
                false,
                applicationInfo,
                null,
                packageInfo.firstInstallTime,
                packageInfo.lastUpdateTime,
                labelResolved,
            )
        }

        private fun configuredPackagesMissingFromCatalog(
            configuredPackages: Set<String>,
            catalog: List<InstalledAppCatalogItem>,
        ): List<String> {
            if (configuredPackages.isEmpty()) return emptyList()
            val installedPackages = catalog.mapTo(HashSet()) { it.packageName }
            return configuredPackages.filterNotTo(ArrayList()) { it in installedPackages }.sorted()
        }

        /**
         * Validates only candidates recovered from persistent state, avoiding store reads for
         * every ordinary installed application.
         */
        @JvmStatic
        fun userVisibleConfiguredPackages(store: DpisConfigStore?): Set<String> {
            if (store == null) return emptySet()
            return store.getConfiguredPackages().filterNotNull().toSet()
                .filterTo(HashSet()) {
                    !SystemFrameworkScope.isFrameworkScopePackage(it) &&
                        store.hasUserVisiblePackageConfig(it)
                }
        }

        /**
         * Keeps scope-only packages aligned with the home configured-app count.
         * A known scope is user-visible configured state even when no package
         * field has been persisted yet.
         */
        @JvmStatic
        fun userVisibleConfiguredPackages(
            store: DpisConfigStore?,
            scopePackages: Set<String>?,
            scopeKnown: Boolean,
        ): Set<String> {
            val configured = userVisibleConfiguredPackages(store).toMutableSet()
            if (scopeKnown) {
                scopePackages.orEmpty()
                    .filterNotTo(configured) { SystemFrameworkScope.isFrameworkScopePackage(it) }
            }
            return configured
        }

        @JvmStatic
        fun countUserVisibleConfiguredPackages(
            store: DpisConfigStore?,
            scopeState: ScopeState?,
        ): Int {
            val safeScopeState = scopeState ?: ScopeState(emptySet(), false)
            return userVisibleConfiguredPackages(
                store,
                safeScopeState.packages,
                safeScopeState.known,
            ).size
        }

        /** An unconfigured row has no persisted package state to decode. */
        @JvmStatic
        fun createUnconfiguredAppListItem(
            label: String,
            packageName: String,
            inScope: Boolean,
            scopeKnown: Boolean,
            systemApp: Boolean,
            hyperOsNativeProxyCandidate: Boolean,
            installed: Boolean,
        ): AppListItem = AppListItem(
            label,
            packageName,
            inScope,
            scopeKnown,
            null,
            null,
            ViewportApplyMode.OFF,
            ViewportTargetType.OFF,
            ViewportTargetSpec.off(),
            null,
            FontApplyMode.OFF,
            null,
            false,
            null,
            true,
            isUserVisibleConfiguredPackage(
                null,
                packageName,
                scopeKnown,
                inScope,
            ),
            installed,
            systemApp,
            hyperOsNativeProxyCandidate,
            null,
        )

        @JvmStatic
        fun createAppListItem(
            store: DpisConfigStore?,
            scopePackages: Set<String>?,
            scopeKnown: Boolean,
            label: String,
            packageName: String,
            systemApp: Boolean,
            hyperOsNativeProxyCandidate: Boolean,
            installed: Boolean,
            icon: Drawable?,
        ): AppListItem {
            val viewportWidth = store?.getTargetViewportWidthDp(packageName)
            val viewportScale = store?.getTargetViewportScaleMilliPercent(packageName)
            val targetSpec = store?.getTargetViewportSpec(packageName) ?: ViewportTargetSpec.off()
            val targetType = store?.getTargetViewportType(packageName) ?: ViewportTargetType.OFF
            val viewportMode = store?.getTargetViewportApplyMode(packageName) ?: ViewportApplyMode.OFF
            val fontScale = store?.getTargetFontScalePercent(packageName)
            val fontMode = store?.getTargetFontApplyMode(packageName) ?: FontApplyMode.OFF
            val typefaceId = store?.getTargetTypefaceId(packageName)
            val fontHookDomainsRaw = store?.getPackageFontHookDomainsRaw(packageName)
            val appSpecificConfigActive = store?.hasTargetAppSpecificConfig(packageName) == true
            val wechatDpi = store?.getWechatDpi(packageName)
            val dpisEnabled = store?.isTargetDpisEnabled(packageName) != false
            val inScope = scopePackages?.contains(packageName) == true
            val configured = isUserVisibleConfiguredPackage(
                store,
                packageName,
                scopeKnown,
                inScope,
            )
            return AppListItem(
                label,
                packageName,
                inScope,
                scopeKnown,
                viewportWidth,
                viewportScale,
                viewportMode,
                targetType,
                targetSpec,
                fontScale,
                fontMode,
                typefaceId,
                appSpecificConfigActive,
                wechatDpi,
                dpisEnabled,
                configured,
                installed,
                systemApp,
                hyperOsNativeProxyCandidate,
                fontHookDomainsRaw,
                icon,
            )
        }

        @JvmStatic
        fun isUserVisibleConfiguredPackage(
            store: DpisConfigStore?,
            packageName: String,
            scopeKnown: Boolean,
            inScope: Boolean,
        ): Boolean =
            !SystemFrameworkScope.isFrameworkScopePackage(packageName) &&
                ((scopeKnown && inScope) || store?.hasUserVisiblePackageConfig(packageName) == true)

        @JvmStatic
        fun shouldUseLauncherVisibilityFallback(
            packageNames: Collection<String>?,
            selfPackageName: String,
        ): Boolean = packageNames.isNullOrEmpty() || packageNames.none {
            it != selfPackageName
        }

        @JvmStatic
        fun isInstalledCatalogChangeAction(action: String?): Boolean = when (action) {
            Intent.ACTION_PACKAGE_ADDED,
            Intent.ACTION_PACKAGE_REMOVED,
            Intent.ACTION_PACKAGE_CHANGED,
            Intent.ACTION_PACKAGE_REPLACED,
            Intent.ACTION_EXTERNAL_APPLICATIONS_AVAILABLE,
            Intent.ACTION_EXTERNAL_APPLICATIONS_UNAVAILABLE,
            -> true
            else -> false
        }

        @JvmStatic
        fun shouldInvalidateInstalledCatalog(action: String?): Boolean =
            isInstalledCatalogChangeAction(action) || action == Intent.ACTION_LOCALE_CHANGED

        private fun queryInstalledPackages(packageManager: PackageManager): List<PackageInfo> =
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                packageManager.getInstalledPackages(PackageManager.PackageInfoFlags.of(0L))
            } else {
                @Suppress("DEPRECATION")
                packageManager.getInstalledPackages(0)
            }

        private fun mergeLauncherPackages(
            packageManager: PackageManager,
            installedPackages: List<PackageInfo>,
        ): List<PackageInfo> {
            val merged = LinkedHashMap<String, PackageInfo>()
            installedPackages.forEach { packageInfo ->
                merged[packageInfo.packageName] = packageInfo
            }
            val launcherIntent = Intent(Intent.ACTION_MAIN).addCategory(Intent.CATEGORY_LAUNCHER)
            val launcherActivities: List<ResolveInfo> =
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    packageManager.queryIntentActivities(
                        launcherIntent,
                        PackageManager.ResolveInfoFlags.of(0L),
                    )
                } else {
                    @Suppress("DEPRECATION")
                    packageManager.queryIntentActivities(launcherIntent, 0)
                }
            launcherActivities.forEach { resolveInfo ->
                val application = resolveInfo.activityInfo?.applicationInfo ?: return@forEach
                if (merged.containsKey(application.packageName)) return@forEach
                merged[application.packageName] = packageInfoForLauncherFallback(
                    packageManager,
                    application,
                )
            }
            return merged.values.toList()
        }

        private fun packageInfoForLauncherFallback(
            packageManager: PackageManager,
            application: ApplicationInfo,
        ): PackageInfo {
            val queried = runCatching {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    packageManager.getPackageInfo(
                        application.packageName,
                        PackageManager.PackageInfoFlags.of(0L),
                    )
                } else {
                    @Suppress("DEPRECATION")
                    packageManager.getPackageInfo(application.packageName, 0)
                }
            }.getOrNull()
            if (queried != null) return queried
            return PackageInfo().also {
                it.packageName = application.packageName
                it.applicationInfo = application
            }
        }

        private fun loadApplicationIcon(
            packageManager: PackageManager,
            applicationInfo: ApplicationInfo,
        ): Drawable? = try {
            applicationInfo.loadIcon(packageManager)
        } catch (_: RuntimeException) {
            null
        }

        private fun loadItemIcon(
            packageManager: PackageManager,
            item: InstalledAppCatalogItem,
        ): Drawable? {
            item.icon?.let { return it }
            return loadApplicationIcon(packageManager, item.applicationInfo)?.also { item.icon = it }
        }

        private fun isSystemApp(applicationInfo: ApplicationInfo): Boolean =
            applicationInfo.flags and ApplicationInfo.FLAG_SYSTEM != 0 &&
                applicationInfo.flags and ApplicationInfo.FLAG_UPDATED_SYSTEM_APP == 0
    }
}
